### PR TITLE
Cli namespaces

### PR DESCRIPTION
**Pull request by frolovdev**
So for now user can pass options to get different schemas and asts divided by different namesapaces

### CODE REVIEW

**Comment by frolovdev** [comment]
concatAST closes #58

**Comment by frolovdev** [comment]
cli code closes #36

File: src/cli/bin.ts [hunk 1]
```diff
@@ -55,36 +108,54 @@ async function main() {
...
-  processingSpinner.succeed();
+  const namespaceSources = namespacesRegexps.map((regexp) =>
+    sources.filter((s) => regexp.test(s.name)),
+  );
+
+  return namespaceSources.map((sourceArray) => sourceArray.concat(commonSources));
 }
 
-main();
+function getGroupedDocuments(groupedSources: Source[][], errCb: (err: Error) => void) {
```
Comment (frolovdev) [inline src/cli/bin.ts:137]: @aimmlegate  please review this code carefully and watch original source codes. IDK maybe I do some typos in mapping of 2 dimensional arrays

File: src/cli/bin.ts [hunk 2]
```diff
@@ -55,36 +108,54 @@ async function main() {
...
+}
 
-  if (parsingError.length > 0) {
-    for (const e of parsingError) {
-      console.error(printCliError(printError(e)));
-    }
-  } else {
-    const doc = { kind: ASTNodeKind.DOCUMENT, definitions: resultDefinitions };
-    const schema = new AnySpecSchema({ ast: doc });
+function groupSourcesByNamespaces({
```
Comment (frolovdev) [inline src/cli/bin.ts:114]: and also this

File: src/cli/bin.ts [hunk 3]
```diff
@@ -55,36 +108,54 @@ async function main() {
...
 
-main();
+function getGroupedDocuments(groupedSources: Source[][], errCb: (err: Error) => void) {
+  const groupedParsedDocuments: DocumentNode[][] = [];
+
+  const parsingErrors: AnySpecError[] = [];
+  groupedSources.forEach((gS) => {
+    const parsedDocuments: DocumentNode[] = [];
+
+    gS.forEach((s) => {
```
Comment (aimmlegate) [inline src/cli/bin.ts:144]: Looks alright but the naming is a bit unclear and strange
let's do smth like this

```
function getGroupedDocuments({
  groupedSources,
  unexpectedErrorHandler,
}: {
  groupedSources: Source[][];
  unexpectedErrorHandler: (err: Error) => void;
}) {
  const groupedParsedDocuments: DocumentNode[][] = [];

  const parsingErrors: AnySpecError[] = [];
  groupedSources.forEach((sourceArray) => {
    const parsedDocuments: DocumentNode[] = [];

    sourceArray.forEach((source) => {
      try {
        const doc = parse(source);
        parsedDocuments.push(doc);
      } catch (error) {
        if (error instanceof AnySpecError) {
          parsingErrors.push(error);
        } else {
          unexpectedErrorHandler(error);
        }
      }
    });
    groupedParsedDocuments.push(parsedDocuments);
  });

  return { groupedParsedDocuments, parsingErrors };
}
```

File: src/cli/bin.ts [hunk 4]
```diff
@@ -3,47 +3,100 @@
...
   const argumentPath = argPaths[0];
   const processingSpinner = ora(`Processing spec: ${argumentPath}`).start();
 
   const specFilePaths = glob(`${argumentPath}/**/*.tinyspec`);
 
+  const sources = await mapPathsToSources(specFilePaths);
+
+  const groupedSources = groupSourcesByNamespaces({ sources, commonNamespace, namespaces });
+
+  const { groupedParsedDocuments, parsingErrors } = getGroupedDocuments(
```
Comment (aimmlegate) [inline src/cli/bin.ts:53]: Maybe it is a good case to use destructured object as a parameter here

smth like this
```
  const { groupedParsedDocuments, parsingErrors } = getGroupedDocuments({
    groupedSources,
    unexpectedErrorHandler: (error: Error) => {
      console.error('Unknown error during parsing', error);
      processingSpinner.fail();
      process.exit(1);
    },
  });
```

File: src/language/concatAST.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { ASTNodeKind, DocumentNode, TypeDefinitionNode } from '../language/ast';
+
+/**
+ * Provided a collection of ASTs, presumably each from different files,
+ * concatenate the ASTs together into batched AST, useful for validating many
+ * GraphQL source files which together represent one conceptual application.
```
Comment (aimmlegate) [inline src/language/concatAST.ts:6]: I think we should get rid of GraphQL mentions in our code 😅

File: src/cli/bin.ts [hunk 6]
```diff
@@ -3,47 +3,100 @@
...
+    }
+
+    processingSpinner.fail();
+    return;
+  }
+
+  const unitedASTs = groupedParsedDocuments.map((documents) => concatAST(documents));
+  const schemas = unitedASTs.map((ast) => new AnySpecSchema({ ast }));
+  const errors = schemas.map((s, index) => validate(s, unitedASTs[index], baseRules));
+  errors.flat().forEach((e) => console.error(printCliError(printError(e))));
```
Comment (aimmlegate) [inline src/cli/bin.ts:74]: Maybe we need to group validation errors by namespaces when we print and separate them from parsing errors in the future?🤔

**Review by aimmlegate — CHANGES_REQUESTED** [review]

File: src/cli/bin.ts (hunk 4 above)
Comment (frolovdev) [reply src/cli/bin.ts:53]: 2 arguments is pretty good for simple parameters

File: src/language/concatAST.ts (hunk 5 above)
Comment (frolovdev) [reply src/language/concatAST.ts:6]: Hm

File: src/cli/bin.ts (hunk 6 above)
Comment (frolovdev) [reply src/cli/bin.ts:74]: For end user it's the same erros

**Pull request merged**